


# promise

```js
let promise = new Promise(function(resolve, reject) {
    // executor
})
```

executor立即执行

```js
let promise = new Promise(function(resolve, reject) {
    resolve("done");
    reject(new Error("..."));
    setTimeout(() => resolve("..."));
});

promise;
/*
Promise {<resolved>: "done"}
    __proto__: Promise
        catch: ƒ catch()
        constructor: ƒ Promise()
            all: ƒ all()
            allSettled: ƒ allSettled()
            arguments: (...)
            caller: (...)
            length: 1
            name: "Promise"
            prototype: Promise {Symbol(Symbol.toStringTag): "Promise", constructor: ƒ, then: ƒ, catch: ƒ, finally: ƒ}
                catch: ƒ catch()
                constructor: ƒ Promise()
                finally: ƒ finally()
                then: ƒ then()
                Symbol(Symbol.toStringTag): "Promise"
                __proto__: Object
            race: ƒ race()
            reject: ƒ reject()
            resolve: ƒ resolve()
            Symbol(Symbol.species): (...)
            get Symbol(Symbol.species): ƒ [Symbol.species]()
            __proto__: ƒ ()
            [[Scopes]]: Scopes[0]
        finally: ƒ finally()
        then: ƒ then()
        Symbol(Symbol.toStringTag): "Promise"
        __proto__: Object
    [[PromiseStatus]]: "resolved"
    [[PromiseValue]]: "done"
*/

promise.__proto__ === Promise.prototype; // true
Promise.prototype.constructor.prototype === Promise.prototype; // true
```

executor 获得结果后调用回调

- 执行成功调用 resolve(value)
- 执行失败调用 reject(error)

promise对象

![image](https://cdn.nlark.com/yuque/0/2020/png/419446/1590160843470-dec89e75-90f3-4075-9cd2-e6e71d9b8524.png)


对比初始状态下的promise
```js
let promise = new Promise(function(resolve, reject) {});
promise;
/*
Promise {<pending>}
    __proto__: Promise
    [[PromiseStatus]]: "pending"
    [[PromiseValue]]: undefined
*/
```

成功返回
```js
let promise = new Promise(function(resolve, reject) {
    setTimeout(() => resolve("done"), 1000);
});

promise;
/*
Promise {<resolved>: "done"}
    __proto__: Promise
    [[PromiseStatus]]: "resolved"
    [[PromiseValue]]: "done"
*/
```

错误处理
```js
let promise = new Promise(function(resolve, reject) {
    setTimeout(() => reject(new Error("Whoops!")), 1000);
});

// Uncaught (in promise) Error: Whoops!

promise;
/*
Promise{<rejected>: Error: Whoops!
    at <anonymous>:2:29}
    __proto__: Promise
    [[PromiseStatus]]: "rejected"
    [[PromiseValue]]: Error: Whoops!
    at <anonymous>:2:29
*/
```

executor只能调用一个结果，成功或错误



## 消费者 then, catch, finally

### then

接收两个参数，两个参数都是函数。第一个函数在resolved后运行并接收结果，第二个函数在rejected后运行并接收error

```js
promise.then(
    function(result) {},
    function(error) {}
    );
```

```js
let promise = new Promise(function(resolve, reject) {
    setTimeout(() => resolve("done!"), 1000);
});

promise.then(
    result => alert(result),
    error => alert(error)
);
```

### catch

error

```js
let promise = new Promise(function(resolve, reject) {
    setTimeout(() => reject(new Error("Wooops")), 1000);
});

promise.catch(alert); // promise.then(null, f)
```

### finally

将结果和error传递给下一个程序
```js
new Promise((resolve, reject) => {
    setTimeout(() => resolve("result"), 2000)
})
.finally(() => alert("Promise ready"))
.then(result => alert(result));
```
error
```js
new Promise((resolve, reject) => {
  throw new Error("error");
})
  .finally(() => alert("Promise ready"))
  .catch(err => alert(err)); 
  
  
new Promise((resolve, reject) => {
    setTimeout(() => reject(new Error("error")), 2000)
})
.finally(() => alert("Promise ready"))
.catch(alert);
```


## 练习

### 基于 promise 的延时
内建函数 setTimeout 使用了回调函数。请创建一个基于 promise 的替代方案。

函数 delay(ms) 应该返回一个 promise。这个 promise 应该在 ms 毫秒后被 resolve，所以我们可以向其中添加 .then，像这样：
```js
function delay(ms) {
  // 你的代码
}

delay(3000).then(() => alert('runs after 3 seconds'));
```

#### 我的解决方案

```js
function delay(ms) {
    return new Promise(function(resolve, reject) {
        setTimeout(() => resolve(), ms);
    });
}

delay(3000).then(() => alert('runs after 3seconds'));
```

#### 官网解决方案

箭头函数

```js
function delay(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}

delay(3000).then(() => alert('runs after 3 seconds'));
```


### 带回调的圆形
改写callback

```js
<!DOCTYPE html>
<html>

<head>
  <meta charset="utf-8">
  <style>
    .message-ball {
      font-size: 20px;
      line-height: 200px;
      text-align: center;
    }
    .circle {
      transition-property: width, height, margin-left, margin-top;
      transition-duration: 2s;
      position: fixed;
      transform: translateX(-50%) translateY(-50%);
      background-color: red;
      border-radius: 50%;
    }
  </style>
</head>

<body>
  <button onclick="go()">Click me</button>

  <script>
  function go() {
    showCircle(150, 150, 100).then(div => {
    div.classList.add('message-ball');
    div.append("Hello, world!");
  });
  }

  function showCircle(cx, cy, radius) {   
    return new Promise(resolve => {
    let div = document.createElement('div');
    div.style.width = 0;
    div.style.height = 0;
    div.style.left = cx + 'px';
    div.style.top = cy + 'px';
    div.className = 'circle';
    document.body.append(div);
    
    setTimeout(() => {
      div.style.width = radius * 2 + 'px';
      div.style.height = radius * 2 + 'px';

      div.addEventListener('transitionend', function handler() {
        div.removeEventListener('transitionend', handler);
        resolve(div);
      });
    }, 0);
    });
  }
  </script>
</body>
</html>
```


# 20200524 手写promise


promise.js
```js
const PENDING = 'PENDING';
const RESOLVED = 'RESOLVED';
const REJECTED = 'REJECTED';

class Promise {
    constructor(executor) {
        this.status = PENDING;
        this.value = undefined;
        this.reason = undefined;

        let resolve = (value) => {
            if (this.status === PENDING) {
                this.value = value;
                this.status = RESOLVED;
                
            }
        }

        let reject = (reason) => {
            if (this.status === PENDING) {
                this.reason = reason;
                this.status = REJECTED;
                
            }
        }

        try {
            executor(resolve, reject);
        }catch(e) {
            reject(e);
        }              
    }

    then(onFulfilled, onRejected) {
        if (this.status === RESOLVED) {
            onFulfilled(this.value);
        }
        if(this.status === REJECTED) {
            onRejected(this.reason);
        }
    }
}

module.exports = Promise;
```

同级目录下 promiseExample.js
```js
let Promise = require('./promise');
let promise = new Promise((resolve, reject) => {
    resolve('成功');
    reject('失败');
});

promise.then(
   (data) => {
       console.log('success', data)
   },
   (err) => {
       console.log('failed', err)
   }
);

// success, 成功
// failed, 失败
```


